data-donation-sdk/src/commonMain/kotlin/care/data4life/datadonation/donation/fhir/anonymization/FhirAnonymizer.kt

```diff
@@ -0,0 +1,87 @@
+ * including any intellectual property rights that subsist in the SDK.
+ *
+ * The SDK and its documentation may be accessed and used for viewing/review purposes only.
+ * Any usage of the SDK for other purposes, including usage for the development of
+ * applications/third-party applications shall require the conclusion of a license agreement
+ * between you and D4L.
+ *
+ * If you are interested in licensing the SDK for your own applications/third-party
+ * applications and/or if you’d like to contribute to the development of the SDK, please
+ * contact D4L by email to help@data4life.care.
+ */
+
+package care.data4life.datadonation.donation.fhir.anonymization
+
+import care.data4life.datadonation.donation.fhir.AllowedReference
+import care.data4life.datadonation.donation.program.model.ProgramAnonymizationGlobalBlur
+import care.data4life.datadonation.donation.program.model.ProgramFhirResourceBlur
+import care.data4life.datadonation.donation.program.model.ProgramType
+import care.data4life.hl7.fhir.stu3.model.FhirResource
+import care.data4life.hl7.fhir.stu3.model.QuestionnaireResponse
+import care.data4life.hl7.fhir.stu3.model.ResearchSubject
+
+internal class FhirAnonymizer(
+    private val blurResolver: AnonymizationContract.BlurRuleResolver,
+    private val questionnaireResponseAnonymizer: AnonymizationContract.QuestionnaireResponseAnonymizer,
+    private val researchSubjectAnonymizer: AnonymizationContract.ResearchSubjectAnonymizer
+) : AnonymizationContract.FhirAnonymizer {
+    private fun anonymizeQuestionaireResponse(
+        questionnaireResponse: QuestionnaireResponse,
+        programType: ProgramType,
+        globalProgramRule: ProgramAnonymizationGlobalBlur?,
+        localResourceRule: Map<AllowedReference, ProgramFhirResourceBlur?>
+    ): QuestionnaireResponse {
+        val rule = blurResolver.resolveBlurRule(
+            fhirResource = questionnaireResponse,
+            globalProgramRule = globalProgramRule,
+            localResourceRule = localResourceRule
+        )
+
+        return questionnaireResponseAnonymizer.anonymize(
+            questionnaireResponse,
+            programType,
+            rule
+        )
+    }
+
+    private fun anonymizeResearchSubject(
+        researchSubject: ResearchSubject,
+        globalProgramRule: ProgramAnonymizationGlobalBlur?
+    ): ResearchSubject {
+        val rule = blurResolver.resolveBlurRule(
+            fhirResource = null,
+            globalProgramRule = globalProgramRule,
+            localResourceRule = emptyMap()
+        )
+
+        return researchSubjectAnonymizer.anonymize(
+            researchSubject,
+            rule
+        )
+    }
+
+    override fun anonymize(
+        fhirResource: FhirResource,
+        programType: ProgramType,
+        globalProgramRule: ProgramAnonymizationGlobalBlur?,
+        localResourceRule: Map<AllowedReference, ProgramFhirResourceBlur?>
+    ): FhirResource {
```

Comment:
same here, maybe might make sense to create an enum type to generalize the anonymize call like i suggested in the validator one for the different type of resources? so you can pass exactly what you need?

Reply:
As said above...I am not getting how a enum can add a benefit here and who is responsible to create it.

Reply:
Also this a [facade](https://en.wikipedia.org/wiki/Facade_pattern).

Reply:
check slack for what i mean in terms of associated Value enum. Facade is not the issue, the issue is its forcing a facade with a non-scalable api